### DISCUSSION TITLE
Thoughts on Avatar Ride-ables

### DISCUSSION THREAD

**Discussion by dpogue** [H-uru/Plasma#929] [Ideas & Proposals]
With the recent release of design docs for Ages like Dahtahmnay and Kahlo, as well as some demos and old source code bits for Kahlo, there is lots of chatter about implementing some of these missing features in the Plasma engine. Most notably, the ability for the avatar to ride on critters/sleds/etc.

### Some Background

The source code for the plAvCritterBrain class used in the Kahlo demo was released, but isn't in a state that's compatible with the current engine codebase. Moreso, plAvCritterBrain was reworked in the current engine to power the Quabs in Ahnonay.
In the Kahlo demo, the avatar model is the critter and human as a single object, with its own avatar animations. It is not an avatar attached to a flying object.

The PotS-era engine had a plVehicleModifier class that allowed for drivable vehicles (powered by Havok physics) which was used for the ZandiMobile. This broke and was removed when MOUL was updated to PhysX in 2006. It appears that newer versions of PhysX support some sort of vehicle stuff, so this could be reimplemented, but is probably tangential to this discussion.
I'm mentioning this just to provide background as to what "rideables" existed and how they differed.

### Use cases for rideables

Rideables are cases where the avatar input controls are handed off to allow controlling some sort of vessel that moves the avatar.

* Flying:
  * Kahlo Kamkentas
  * Hang-gliders
  * Parachutes
* Drivables:
  * Boats
  * Sleds (Maybe plVehicleModifier?)
  * Possibly simple vehicles (For real vehicles we'd probably want to reintroduce plVehicleModifier)

### Requirements

* Avatar can "mount" and "dismount" the rideable and controls behave normally when not riding
* Avatar can either be attached to the rideable or hidden, while riding
* Avatar can play an animation while riding
  * If we want to get really fancy, avatar can have multiple animations for different states
* Rideable can have animations for different states (which probably implies being plArmatureMods)
* Rideables can have different gravity (for flying)
* Do **not** want a special case C++ brain class for every rideable use case
  * This means configuration for animations and parameters like maximum speed and gravity probably need to be serialized in as a new type of plModifier

### Misc Brainstorming
* Mount/dismount regions, so that you can only get on/off at certain locations (like a boat at docks)
* Some rideables might not need animations and can be simple objects (again, boats)
* Are custom controls needed for different rideables?
* First/Third person cameras?
* Definitely need some Python API for this, TBD what is configurable there and what the methods would be

**Comment by dpogue** [comment]
Thinking through the details of this a bit more, we probably need every rideable to be a `plArmatureMod` for ease of using the physical controller. We might be able to reuse `plAvCritterBrain`, since it doesn't actually do much behaviour itself and is exposed to Python for scripting (which is [how the Quabs are handled](https://github.com/H-uru/Plasma/blob/master/Scripts/Python/ahnyQuabs.py) today).

`plAvBrainRideAnimatedPhysical` was originally intended for Jalak-style animated physicals, but that's actually all handled by `plAvBrainHuman` and `plWalkingStrategy` now, leaving `plAvBrainRideAnimatedPhysical` entirely unused. So we can probably modify that to do whatever we need rather than adding a new brain.

My current thinking is something like push a `plAvBrainRideAnimatedPhysical` onto the player also give it the `plAvCritterBrain` of the rideable, and then allow it to pass messages/control down to the critter brain.

**Comment by Emor-Dni-Lap** [comment]
Agree with everything you've written here, dpogue; it all makes sense.
Wish I could assist with moving any of this forward, but it's out of my league! :)

I can only offer a few additional thoughts to amplify what you've already said, specifically regarding Kahlo:

Requirements
- Avatar mounting and unmounting will in some cases require new animations. We will need to get better at this (yes, that includes me)
However, I'd guess that Cyan's intention was to do a fade-out as soon as a critter (Kamkenta?) was selected in the upper pod, then a fade-up on the united critter+suited-up (and thereby non-unique) mounted-avatar combo on the playing field....avoiding a mounting animation altogether.
- The flying combo will need appropriate cycling animations for all its dynamic states (i.e. when a flying vehicle is doing a banking maneuver or in steep ascent or descent, specific avatar animations are switched in). Of course, this would require a feature to set envelope parameters for these triggering-points (LUT?).
- Flying vehicles should have a settable constant for a banking parameter tied to L/R yaw rotation: the harder the turn, the more Z-rotation banking on the rideable
- Would also be nice for rideables to have a parameter for random and/or programmed buffeting (wind, irregular ground, etc.)
- And yes: envelopes for max speed, min/max elevation above groundplane, min/max slope angle (climb/descent) and max banking for flying vehicles. End settings on these envelopes should also have a parameter for damping, so that limits are never hard-hit. 

Seems like, if enough of these options could be switchable (some must remain unswitchable!), settings would be flexible enough to cover most rideables I can think of, without requiring custom controls for each type. Since there might be quite a lot of variables here though, saveable template creation could be handy. Just pipe-dreaming... sorry! :D
